### PR TITLE
statistics: support tracking histogram

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/34052

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [x] Manual test (add detailed scripts or steps below)

[test report](https://pingcap.feishu.cn/docx/doxcnZ0e75oMBm959tPu7NL8Dag)

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: statistics/handle/lru_cache.go [hunk 1]
```diff
@@ -430,7 +430,7 @@ func (c *innerItemLruCache) put(tblID, id int64, isIndex bool, item statistics.T
 
 func (c *innerItemLruCache) evictIfNeeded() {
 	curr := c.cache.Back()
-	for c.trackingCost > c.capacity {
+	for c.trackingCost > c.capacity && curr != nil {
```
Comment (chrysan) [inline statistics/handle/lru_cache.go:433]: When curr is nil?

File: statistics/handle/lru_cache.go [hunk 2]
```diff
@@ -443,7 +443,7 @@ func (c *innerItemLruCache) evictIfNeeded() {
 			c.cache.Remove(curr)
 			delete(c.elements[item.tblID][item.isIndex], item.id)
 		} else {
-			c.cache.PushFront(curr)
+			c.cache.MoveToFront(curr)
```
Comment (chrysan) [inline statistics/handle/lru_cache.go:446]: Is PushFront a bug?

File: statistics/histogram.go [hunk 3]
```diff
@@ -139,8 +139,10 @@ func (hg *Histogram) MemoryUsage() (sum int64) {
 	if hg == nil {
 		return
 	}
-	sum = EmptyHistogramSize + hg.Bounds.MemoryUsage() + int64(cap(hg.Buckets))*EmptyBucketSize + int64(cap(hg.scalars))*EmptyScalarSize
-	return
+	if len(hg.Buckets) == 0 && len(hg.scalars) == 0 && hg.Bounds.Capacity() == 0 {
+		return
+	}
+	return hg.Bounds.MemoryUsage()
```
Comment (chrysan) [inline statistics/histogram.go:145]: What's the purpose of this change?

Comment (Yisaer) [reply statistics/histogram.go:145]: We found the original calculation has large diff from the test in https://pingcap.feishu.cn/docx/doxcnZ0e75oMBm959tPu7NL8Dag#doxcn4QOMAuk6icayQX6KmGmY3f

File: statistics/handle/lru_cache.go (hunk 1 above)
Comment (Yisaer) [reply statistics/handle/lru_cache.go:433]: In fact, it won't become nil

File: statistics/column.go [hunk 4]
```diff
@@ -346,6 +346,13 @@ func (c *Column) dropTopN() {
 	}
 }
 
+func (c *Column) dropHist() {
+	c.Histogram.Bounds = chunk.NewChunkWithCapacity([]*types.FieldType{types.NewFieldType(mysql.TypeBlob)}, 0)
+	c.Histogram.Buckets = make([]Bucket, 0)
+	c.Histogram.scalars = make([]scalar, 0)
```
Comment (xuyifangreeneyes) [inline statistics/column.go:352]: Can we just set `Bounds`, `Buckets` and `scalars` to nil?

File: statistics/histogram.go [hunk 5]
```diff
@@ -139,8 +139,11 @@ func (hg *Histogram) MemoryUsage() (sum int64) {
 	if hg == nil {
 		return
 	}
+	if len(hg.Buckets) == 0 && len(hg.scalars) == 0 && hg.Bounds.Capacity() == 0 {
+		return
```
Comment (xuyifangreeneyes) [inline statistics/histogram.go:143]: `hg` still takes `EmptyHistogramSize` if it is not nil?

File: statistics/histogram.go [hunk 6]
```diff
@@ -1591,6 +1594,11 @@ func (s StatsLoadedStatus) IsTopNEvicted() bool {
 	return s.statsInitialized && s.evictedStatus >= onlyHistRemained
 }
 
+// IsAllEvicted  indicates whether all the stats got evicted not.
```
Comment (xuyifangreeneyes) [inline statistics/histogram.go:1597]:
```suggestion
// IsAllEvicted indicates whether all the stats got evicted or not.
```

File: statistics/column.go (hunk 4 above)
Comment (Yisaer) [reply statistics/column.go:352]: This may cause panic in some Histogram functions which are hard to cover.

File: statistics/histogram.go (hunk 5 above)
Comment (Yisaer) [reply statistics/histogram.go:143]: Right, but we will think it as 0, the difference is acceptable.

**Review by xuyifangreeneyes — APPROVED** [review]

File: statistics/histogram.go [hunk 7]
```diff
@@ -139,8 +139,11 @@ func (hg *Histogram) MemoryUsage() (sum int64) {
 	if hg == nil {
 		return
 	}
+	if len(hg.Buckets) == 0 && len(hg.scalars) == 0 && hg.Bounds.Capacity() == 0 {
+		return
+	}
```
Comment (chrysan) [inline statistics/histogram.go:144]: Empty histogram consumes memory. Is there a way to track and evict the empty part too?

Comment (Yisaer) [reply statistics/histogram.go:144]: We use empty histograms in order to avoid panic problem. Thinking empty histograms as 0 would be friendly to our lru cache as finally tracking cost can be 0. I think the difference of the memory cost for empty histogram and zero is accptable.

Comment (chrysan) [reply statistics/histogram.go:144]: It's fine for most of the cases, while for some cases with many tables(more than 100,000), let's consider it in another topic.

**Review by chrysan — APPROVED** [review]
